default/AI/ShipDesignAI.py

```diff
@@ -1442,6 +1479,51 @@ def _partclass_in_design(self, partclass):
         """
         return any(part.partClass in partclass for part in self.parts)
 
+    def _calculate_weapon_strength(self, weapon):
+        # base damage
+        weapon_name = weapon.name
+        damage = weapon.capacity
+        # species modifier
+        weapons_grade, _, _ = CombatRatingsAI.get_piloting_grades(self.species)
+        species_modifier = AIDependencies.PILOT_DAMAGE_MODIFIER_DICT.get(weapons_grade, {}).get(weapon_name, 0)
+        # tech upgrades
+        try:
+            upgrades = AIDependencies.WEAPON_UPGRADE_DICT[weapon_name]
+        except KeyError:
+            global _raised_warnings
+            if weapon_name not in _raised_warnings:
+                _raised_warnings.add(weapon_name)
+                print_error(("WARNING: Encountered unknown weapon (%s): "
+                             "The AI will play on but damage estimates may be incorrect leading to worse performance. "
+                             "Please update AIDependencies.py and "
+                             "add the weapon with its upgrade techs to WEAPON_UPGRADE_DICT") % weapon_name,
+                            location="ShipDesignAI._calculate_weapon_strength()", trace=True)
+            return damage
+        total_tech_bonus = 0
+        for tech, dmg_bonus in upgrades:
+            total_tech_bonus += dmg_bonus if tech_is_complete(tech) else 0
+            # TODO: Error checking if tech is actually a valid tech (tech_is_complete simply returns false)
```

Comment:
May be we should wrap: `upgrades = AIDependencies.WEAPON_UPGRADE_DICT[weapon_name]` to function, that will return valid tech names. 

I don't like to add validation logic that scripts and AI is synced to place where we make decisions.